### PR TITLE
Implement sum for cupy.sparse

### LINKED ISSUE #595: Any function like scipy.sparse.csc_matrix.sum ?
Thanks for sharing this awesome framework. Is there any function similar to [scipy.sparse.csc_matrix.sum](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.sum.html#scipy.sparse.csc_matrix.sum) ?

### PR DESCRIPTION
**Pull request by unnonouno**
This patch includes most general sum implementation ported from `scipy.sparse.spmatrix`. We can implement more efficient one for each matrix type.
fix #595

### CODE REVIEW

**Comment by unnonouno** [comment]
jenkins test this please

File: cupy/sparse/base.py [hunk 1]
```diff
@@ -321,7 +321,52 @@ def set_shape(self, shape):
...
+        """
+        util.validateaxis(axis)
+
+        # This implementation uses multiplication, though it is not efficient
+        # for some matrix types. These should override this function.
+
+        m, n = self.shape
+
+        if axis is None:
+            return (self.dot(cupy.ones(n, dtype=self.dtype))).sum(
```
Comment (okuta) [inline cupy/sparse/base.py]: Please remove extra parenthesis.

File: cupy/sparse/base.py [hunk 2]
```diff
@@ -321,7 +321,52 @@ def set_shape(self, shape):
...
+        else:  # axis == 1
+            ret = self.dot(cupy.ones(n, dtype=self.dtype)).reshape(m, 1)
+
+        if out is not None:
+            if out.shape != ret.shape:
+                raise ValueError('dimensions do not match')
+            out[:] = ret
+            return out
+        elif dtype is not None:
+            return ret.astype(dtype)
```
Comment (okuta) [inline cupy/sparse/base.py]: How about `return ret.astype(dtype, copy=False)`?

**Comment by unnonouno** [comment]
I fixed it

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**